packages/Search/src/details/DTK_DetailsDistributor.hpp

```diff
@@ -0,0 +1,252 @@
+        counts.push_back( 0 );
+        // TODO consider replacing with parallel scan
+        for ( int i = 0; i < n; ++i )
+        {
+            if ( ranks_duplicate( i ) == largest_rank )
+            {
+                ranks_duplicate( i ) = -1;
+                permutation_indices( i ) = offsets.back() + counts.back();
+                ++counts.back();
+            }
+        }
+        offsets.push_back( offsets.back() + counts.back() );
+    }
+}
+
+// A number of MPI routines (in the current OpenMPI implementation at least) are
+// checking at runtime that all pointers passed to them as argument are not null
+// and call program termination even when no data would actually be
+// communicated. This is problematic when using std::vector to store buffers and
+// other arrays because std::vector::data() may return a null pointer if the
+// vector is empty. This function "swaps" the pointer with some other pointer
+// that is not dereferenceable if it is the null pointer.
+template <typename T>
+static T *notNullPtr( T *p )
+{
+    return p ? p : reinterpret_cast<T *>( -1 );
+}
+
+class Distributor
+{
+  public:
+    Distributor( MPI_Comm comm )
+        : _comm( comm )
+    {
+    }
+
+    template <typename View>
+    size_t createFromSends( View const &destination_ranks )
+    {
+        static_assert( View::rank == 1, "" );
+        static_assert(
+            std::is_same<typename View::non_const_value_type, int>::value, "" );
+        int comm_rank;
+        MPI_Comm_rank( _comm, &comm_rank );
+
+        _permute = Kokkos::View<int *, Kokkos::HostSpace>(
+            Kokkos::ViewAllocateWithoutInitializing( "permute" ),
+            destination_ranks.size() );
+        std::vector<int> destinations;
+        sortAndDetermineBufferLayout( destination_ranks, _permute, destinations,
+                                      _dest_counts, _dest_offsets );
+
+        {
+            int const reorder = 0; // Ranking may *not* be reordered
+            int const sources[1] = {comm_rank};
+            int const degrees[1] = {static_cast<int>( destinations.size() )};
+
+            MPI_Dist_graph_create(
+                _comm, 1, sources, degrees, notNullPtr( destinations.data() ),
+                MPI_UNWEIGHTED, MPI_INFO_NULL, reorder, &_comm_dist_graph );
+        }
+
+        int indegrees;
+        int outdegrees;
+        int weighted;
+        MPI_Dist_graph_neighbors_count( _comm_dist_graph, &indegrees,
+                                        &outdegrees, &weighted );
+        DTK_ENSURE( weighted == 0 );
+        DTK_ENSURE( outdegrees == static_cast<int>( destinations.size() ) );
+
+        std::vector<int> sources( indegrees );
+        MPI_Dist_graph_neighbors( _comm_dist_graph, indegrees,
+                                  notNullPtr( sources.data() ), MPI_UNWEIGHTED,
+                                  outdegrees, notNullPtr( destinations.data() ),
+                                  MPI_UNWEIGHTED );
```

Comment:
Is there a guarantee that the neighbors are going to be listed in the same order as original `destinations`? Could not find this anywhere in the documentation. In other words, could this call permute `destinations`? If yes, then the next `MPI_Neighbor_alltoall` may do the wrong thing.

Reply:
> Is there a guarantee that the neighbors are going to be listed in the same order as original `destinations`? 

No

> If yes, then the next `MPI_Neighbor_alltoall` may do the wrong thing.

I don't see how "permute destinations" would make  `MPI_Neighbor_alltoall` do the wrong thing

Reply:
`destination` is an output parameter (I realize this is confusing because it overwrite whatever was in the vector until this point)
This is precisely how MPI tells you how receives are going to be laid off.

Reply:
Talked to @dalg24 about the thing, and he is going to investigate this.